TheJumpCloud/support
# SA-2641/2668: Get-JCCommandResult Parallel and -CommandID implementation

## Issues
* [SA-2668](https://jumpcloud.atlassian.net/browse/SA-2668) - Get-JCCommandResult should be used to get all the results from that specific workflowID
* [SA-2641](https://jumpcloud.atlassian.net/browse/SA-2641) - Get-JCCommandResult - Command ID parameter

## What does this solve?
* Added `-ByCommandID` and `-CommandID` to Get-JCCommandResult
    * The added functionality will allow admins to search for all command results pertaining to a single command via the commandID or the workflowID
    * When using the pipeline for inputting a command object to Get-JCCommandResult, use the -ByCommandID switch
        * Example: `$OrgCommandResults = Get-JCCommand | Get-JCCommandResult -ByCommandID`
    * When using the pipeline for inputting a commandResult object to Get-JCCommandResult, use the -ByID switch
        * Example: `$OrgCommandResults = Get-JCCommandResult | Get-JCCommandResult -ByID`
* Deprecated -Skip, -Limit, -MaxResults parameters from Get-JCCommandResult

## Is there anything particularly tricky?

* Some previous functionality has been removed in favor of a more straight forward workflow
    * `-Skip, -Limit, -MaxResults` have been removed because the function is going to handle those parameters internally without need for user input
* You are no longer able to get the /commandresult/{id} endpoint result by piping Get-JCCommandResult into itself without specifying the -ByID switch
    * Previous functionality: 
        * `$SingleCommandResult = Get-JCCommandResult | Select-Object -Last 1 | Get-JCCommandResult`
    * A workaround to this is to use the -ByID switch
        * `$SingleCommandResult = Get-JCCommandResult | Select-Object -Last 1 | Get-JCCommandResult -ByID`

## How should this be tested?
* New Tests have been written for the `-ByCommandID` and the `-CommandID/-WorkflowID` parameters
```powershell
Describe -Tag:('JCCommandResult') "Get-JCCommandResult 2.0" {
    It "Returns command results with the workflowId" {
        $Command = Get-JCCommandResult | Select-Object -First 1
        $CommandResults = Get-JCCommandResult -WorkflowId $Command.workflowId
        $CommandResults | Should -Not -BeNullOrEmpty
    }
    It "Returns command results with the CommandId" {
        $Command = Get-JCCommandResult | Select-Object -First 1
        $CommandResults = Get-JCCommandResult -CommandID $Command.workflowId
        $CommandResults | Should -Not -BeNullOrEmpty
    }
    It "Returns all results for Command via pipeline (command object)" {
        $CommandResults = Get-JCCommand | Get-JCCommandResult -ByCommandID
        $CommandResults | Should -Not -BeNullOrEmpty
    }
}
```
* All previous tests should pass excluding the `-Skip, -Limit, -MaxResults` tests that have been removed and `Gets a single JumpCloud command result passed through the pipeline without declaring -ByID`
